core/src/main/scala/better/files/Implicits.scala

```diff
@@ -139,29 +135,66 @@ trait Implicits {
     }
 
     /**
-     * Provides and iterator that closes the underlying resource
-     * when done.
-     * 
+     * Provides an iterator that closes the underlying resource when done
+     *
      * e.g.
      * <pre>
      * inputStream.autoClosedIterator(_.read())(_ != -1).map(_.toByte)
      * </pre>
      *
-     * @param next next element in this resource
-     * @param hasNext a function which tells if there is no more B left
+     * @param generator next element from this resource
+     * @param isValidElement a function which tells if there is no more B left e.g. certain iterators may return nulls
      * @tparam B
-     * @return An iterator that closes itself when done
+     * @return An iterator that closes the underlying resource when done
      */
-    def autoClosedIterator[B](next: A => B)(hasNext: B => Boolean): Iterator[B] = {
+    def autoClosedIterator[B](generator: A => B)(isValidElement: B => Boolean): Iterator[B] = {
       var isClosed = false
       def isOpen(item: B) = {
-        if(!isClosed && !hasNext(item)) {
-          resource.close()
-          isClosed = true
-        }
+        if(!isClosed && !isValidElement(item)) close()
         !isClosed
       }
-      Iterator.continually(next(resource)).takeWhile(isOpen)
+
+      def close() = try {
+        resource.close()
+      } finally {
+        isClosed = true
+      }
+
+      def next() = try {
+        generator(resource)
+      } catch {
+        case NonFatal(e) =>
+          close()
+          throw e
+      }
+
+      Iterator.continually(next()).takeWhile(isOpen)
+    }
+  }
+
+  implicit class JStreamOps[A](stream: JStream[A]) {
+    /**
+     * Closes this stream when iteration is complete
+     * It will NOT close the stream if it is not depleted!
+     *
+     * @return
+     */
+    def toAutoClosedIterator: Iterator[A] = {
+      val iterator = stream.iterator()
+      var isClosed = false
+      new Iterator[A] {
+        override def hasNext = {
+          if (!isClosed && !iterator.hasNext) {
+            try {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Consider using util.Try instead of a try-catch block](https://www.codacy.com/app/pathikrit/better-files/file/1938395408/issues/source?bid=2860326&fileBranchId=3103914#l188)